packages/material-ui-lab/src/MasonryItem/MasonryItem.js

```diff
@@ -0,0 +1,222 @@
+  const props = useThemeProps({
+    props: inProps,
+    name: 'MuiMasonryItem',
+  });
+
+  const masonryItemRef = React.useRef(null);
+
+  const { spacing = 1 } = React.useContext(MasonryContext);
+  const isImage = (element) => {
+    return element.type === 'img';
+  };
+  const {
+    children,
+    className,
+    component = 'div',
+    columnSpan = 1,
+    height = isImage(children) ? 1 : undefined,
+    // images should have defined dimensions for browser to correctly lay out and paint.
+    // browser can't handle a large number of images without pre-defined dimensions
+    ...other
+  } = props;
+  const isSSR = !((isImage(children) && height === 1) || height === undefined);
+  const [styleProps, setStyleProps] = React.useState({
+    ...props,
+    isSSR,
+    isImage: isImage(children),
+    spacing,
+    columnSpan,
+    height: height < 0 ? 0 : height, // MasonryItems to which negative or zero height is passed will be hidden
+  });
+
+  const classes = useUtilityClasses(styleProps);
+  const resizeObserver = React.useRef(null);
+
+  React.useEffect(() => {
+    if (isImage(children) && styleProps.height === 1) {
+      const img = new Image();
+      img.src =
+        children.props.srcSet ||
+        children.props.src ||
+        children.props['data-src'] ||
+        children.props['data-srcSet'];
+      img.onload = () => {
+        setStyleProps({
+          ...styleProps,
+          height: img.height,
+        });
+      };
+    }
+  }, [children, styleProps]);
+
+  React.useEffect(() => {
+    // do not create a resize observer in case of SSR masonry
+    if (isSSR) {
+      return true;
+    }
+    // do not create a resize observer until height is set by onload() callback above
+    if (isImage(children) && styleProps.height === 1) {
+      return true;
+    }
+    const handleImageLoadError = () => {
+      const src = masonryItemRef.current.firstChild.src;
+      masonryItemRef.current.firstChild.src = `${src}?${new Date().getTime()}`;
+    };
+    resizeObserver.current = new ResizeObserver(([item]) => {
+      if (styleProps.height !== item.contentRect.height) {
+        setStyleProps({
+          ...styleProps,
+          height: item.contentRect.height,
+        });
+      }
+    });
+    const item = masonryItemRef.current.firstChild;
+    const observer = resizeObserver.current;
+    item.addEventListener('error', handleImageLoadError);
+    observer.observe(item);
+    return () => {
+      observer.unobserve(item);
+      item.removeEventListener('error', handleImageLoadError);
+    };
+  }, [children, isSSR, styleProps]);
+
+  React.useEffect(() => {
+    // handle responsiveness to `height` prop in case of SSR masonry
+    if (isSSR && height !== styleProps.height) {
+      setStyleProps({
+        ...styleProps,
+        height: height < 0 ? 0 : height,
+      });
+    }
+    // re-style if `spacing` updates
+    if (spacing !== styleProps.spacing) {
+      setStyleProps({ ...styleProps, spacing });
```

Comment:
This looks like derived state from props which should not be updated in an effect since it leads to tearing: One commit includes the old `height` and new `styleProps.height`, the next commit includes the new `height` and new `styleProps.height`. You can move this into the body of the function component. Same for the other updates.

Reply:
How was this conversation resolved?

Reply:
I just realised I didn't push my commits. My bad.

Reply:
Moved conversation to https://github.com/hbjORbj/material-ui/pull/1